### ISSUE TITLE
Parity performance problem with Ledger attached/contract usage.

### ISSUE THREAD

**Issue by MicahZoltu** [paritytech/parity#6405]
_Before filing a new issue, please **provide the following information**._

> I'm running:
>
> - **Parity version**: Parity//v1.7.0-beta-5f2cabd-20170727/x86_64-windows-msvc/rustc1.18.0
> - **Operating system**: Windows
> - **And installed**: via installer

_Your issue description goes here below. Try to include **actual** vs. **expected behavior** and **steps to reproduce** the issue._

---

I'm sure you all are tired of me reporting these hard-to-reproduce performance problems, but I seem to be completely plagued with them.  The latest is on my _new_ computer, for those following along.  This computer was previously running without trouble but is now experiencing similar symptoms to my previous computer (though this time there was a trigger).

It started when I plugged in my ledger and started interacting with a contract of mine in the UI (signing with the ledger).  Even after ceasing interactions, my CPU is still running quite hot (though it calms to ~15% and then spikes to ~85% for extended periods of each).  Logs do not show any out-of-the-normal activity.

Unplugging the ledger does not immediately resolve the issue.

Here is a sample of the logs:
```
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎41‎:‎29] Imported #4214901 c5a4…2e15 (145 txs, 6.69 Mgas, 91.71 ms, 20.23 KiB)
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎41‎:‎26] Imported #4214901 1deb…f8ba (141 txs, 6.70 Mgas, 100.19 ms, 18.75 KiB)
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎41‎:‎22] Imported #4214900 b9db…b6d7 (231 txs, 6.69 Mgas, 176.26 ms, 30.15 KiB)
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎41‎:‎13] 20/25 peers 2 MiB chain 64 MiB db 0 bytes queue 23 KiB sync RPC: 1 conn, 6 req/s, 134 µs
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎40‎:‎40] Imported #4214899 55e5…8d31 (121 txs, 3.07 Mgas, 48.88 ms, 15.66 KiB)
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎40‎:‎40] Transaction mined (hash c63683306142d2548b21b130db6af18b90a3a80e54bade48a52680b34aca6e23)
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎40‎:‎38] 23/25 peers 6 MiB chain 64 MiB db 0 bytes queue 23 KiB sync RPC: 1 conn, 11 req/s, 273 µs
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎40‎:‎33] Imported #4214898 2aa8…d0da (121 txs, 6.70 Mgas, 69.27 ms, 18.03 KiB)
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎40‎:‎03] 21/25 peers 6 MiB chain 64 MiB db 0 bytes queue 23 KiB sync RPC: 1 conn, 6 req/s, 267 µs
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎39‎:‎51] Imported #4214897 c148…cfda (183 txs, 6.67 Mgas, 135.63 ms, 29.82 KiB)
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎39‎:‎49] Imported #4214897 0bb0…1530 (138 txs, 6.72 Mgas, 77.23 ms, 19.09 KiB)
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎39‎:‎28] 23/25 peers 6 MiB chain 64 MiB db 0 bytes queue 23 KiB sync RPC: 1 conn, 9 req/s, 11384974 µs
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎39‎:‎28] Imported #4214896 f92b…fd04 (138 txs, 6.71 Mgas, 66.35 ms, 17.39 KiB)
[‎8‎/‎28‎/‎2017‎ ‎19‎:‎39‎:‎12] Imported #4214895 38c2…9436 (83 txs, 6.72 Mgas, 51.10 ms, 13.63 KiB)
```

CPU utilization chart of just parity.exe:  
![image](https://user-images.githubusercontent.com/886059/29802294-120854ac-8c29-11e7-87df-5ab94d4e7cc4.png)

IO is 2-20 MB/s.

Closing the UI does not resolve the issue.  Restarting Parity does resolve the issue temporarily, but it appears to have come back (even though the ledger has been unplugged since I restarted.

Opening the UI and forgetting the ledger appears to have made the problem even worse.  RAM usage continues to rise the longer the problem persists, though I am not using enough to cause any problems on this machine (only ~700MB committed).

Restarting Parity after forgetting my ledger also does not resolve the problem.  It would seem that my Parity node is now in a permanently broken state, starting around the time I plugged in my Ledger and used it to interact with a contract inside Parity UI.

----

I would love it if a developer could walk me through the best way to get some debugging data on Windows.  Having had to debug problems like this in the past, the number one useful thing was a thread profile.  However, I don't know how to best capture a thread profile for a Rust application on Windows.  Please advise how I can capture such a thing so Parity developers can look to see where Parity is burning cycles!

**Comment by MicahZoltu** [comment]
I found a `LOG` file in my feeble attempts to self troubleshoot that appears to indicate that Parity is doing something at incredibly high frequency (and logging all of it).  I don't understand the logs so I can't speak to what it is, but based on the timestamps I would assume that it is not normal to repeat at this speed when sitting idle.

Gist: https://gist.github.com/MicahZoltu/c3241b8d81ce078b564592439126ee89

**Comment by 5chdn** [comment]
Thanks, @MicahZoltu, I really appreciate your detailed reports.

> I'm sure you all are tired of me reporting these hard-to-reproduce performance problems, but I seem to be completely plagued with them. 

You are not alone and we are currently working on multiple performance fixes.

> I would love it if a developer could walk me through the best way to get some debugging data on Windows.

The most simple way to do this would be writing a log file. Please, locate your [config file]https://github.com/paritytech/parity/wiki/Configuring-Parity#config-file) which is probably in:

```bash
%USERPROFILE%\AppData\Roaming\Parity\Ethereum\config.toml
```

Where `%USERPROFILE%` should resolve to `C:\Documents and Settings\YourName`. If it does not exist, please create it with the following contents:

```toml
[parity]
chain = "foundation"

[misc]
logging = "rpc=trace"
log_file = "%TEMP%\Parity\debug.log"
```

This will write the logs to `C:\Documents and Settings\YourName\Local Settings\Temp\Parity\debug.log` and the logging level `rpc=trace` will allow us some insight what the UI is doing while you use it. (Literally, tracing the RPC.)

You can fine tune the log level from trace, debug, info, warn, to error and enable or disbale [more modules](https://ethereum.stackexchange.com/questions/3331/how-to-make-parity-write-logs). But let's have a look at RPC first, because, once again, I suspect it's related to #6387 

> I found a `LOG` file in my feeble attempts to self troubleshoot that appears to indicate that Parity is doing something at incredibly high frequency (and logging all of it).

That's just the RocksDB debug logs, not of any use here.

**Comment by MicahZoltu** [comment]
I'll get that data together.  For future reference though, your instructions could use a bit of improvement.  :)

`%APPDATA%` should be used instead of `%USERPROFILE%\AppData\Roaming` as the folder could move depending on OS setup.

`%TEMP%` by default is `%LOCALAPPDATA%\Temp` which (on modern Windows) is `C:\Users\<username>\AppData\Local\Temp` by default.

Also, `%USERPROFILE%` is `C:\Users\<username>` by default on modern versions of Windows.

----

I'll apply these configuration settings and see if I can still reproduce the problem (Parity has been off all night because of this issue).

**Comment by MicahZoltu** [comment]
```
C:\Program Files\Parity\Parity>parity.exe ui
Loading config file from C:\Users\micah\AppData\Roaming\Parity\Ethereum\config.toml
thread 'main' panicked at 'Logger is initialized only once; qed: "Cannot write to log file given: %TEMP%\\Parity\\debug.log"', src\libcore\result.rs:859
stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: <unknown>
  12: <unknown>
  13: <unknown>
  14: hid_error
  15: BaseThreadInitThunk

C:\Program Files\Parity\Parity>
```

**Comment by MicahZoltu** [comment]
With the UI open, there is definitely a ton of RPC traffic based on the logs.  Howwever, as soon as I closed the UI the logs switched to just the following, yet CPU utilization remained high:
```
2017-08-29 11:00:24  IO Worker #0 INFO import  Syncing #4215852 5e6b…718b     6 blk/s  756 tx/s  34 Mgas/s      0+  644 Qed  #4216496   13/25 peers      6 MiB chain   78 MiB db   62 MiB queue   10 MiB sync  RPC:  0 conn,  8 req/s, 2596 µs
2017-08-29 11:00:35  IO Worker #3 INFO import  Syncing #4215997 1420…3cec    13 blk/s 1408 tx/s  78 Mgas/s      0+  496 Qed  #4216496   14/25 peers      5 MiB chain   79 MiB db   48 MiB queue   10 MiB sync  RPC:  0 conn,  0 req/s, 2596 µs
2017-08-29 11:00:44  IO Worker #3 INFO import  Syncing #4216106 426d…0f3d    11 blk/s 1432 tx/s  69 Mgas/s      0+  388 Qed  #4216496   15/25 peers     11 MiB chain   78 MiB db   36 MiB queue   10 MiB sync  RPC:  0 conn,  0 req/s, 2596 µs
2017-08-29 11:00:54  IO Worker #3 INFO import  Syncing #4216204 1219…f571     9 blk/s 1079 tx/s  54 Mgas/s      0+  954 Qed  #4217158   16/25 peers      7 MiB chain   78 MiB db   86 MiB queue  667 KiB sync  RPC:  0 conn,  0 req/s, 2596 µs
2017-08-29 11:00:59  IO Worker #3 INFO import  Syncing #4216285 de00…4aad    15 blk/s 1394 tx/s  77 Mgas/s      0+  870 Qed  #4217158   14/25 peers      5 MiB chain   78 MiB db   78 MiB queue  667 KiB sync  RPC:  0 conn,  0 req/s, 2596 µs
2017-08-29 11:01:09  IO Worker #2 INFO import  Syncing #4216392 f65a…08e7    11 blk/s 1390 tx/s  69 Mgas/s      0+  762 Qed  #4217158   15/25 peers      6 MiB chain   77 MiB db   66 MiB queue  667 KiB sync  RPC:  0 conn,  0 req/s, 2596 µs
2017-08-29 11:01:15  IO Worker #2 INFO import  Syncing #4216464 f9be…f845    11 blk/s 1005 tx/s  71 Mgas/s      0+  695 Qed  #4217159   14/25 peers      8 MiB chain   78 MiB db   61 MiB queue  667 KiB sync  RPC:  0 conn,  0 req/s, 2596 µs
```

Note: I'm still catching up on syncing, will update once I'm fully synced.  I'll also try different logging options to see if I can find root cause.

**Comment by MicahZoltu** [comment]
Upon fully syncing, the problem appears to have gone away.  As a next step in troubleshooting, I plugged in my Ledger.  Unfortunately, Parity didn't pick it up right away (saw it as in use by another application).  Upon restarting Parity (with the ledger plugged in and Ethereum app open on the Ledger), the problem came back.  The RPC logs show no activity (UI is closed) other than the standard syncing log lines every 15 seconds or so.

Restarting Parity again (Ledger still plugged in and Ethereum app open on it) but this time without the `ui` flag resulted in the problem going away.  Upon opening the UI, CPU burned a bit warm (but not hot) for about a minute and the logs were flooded with RPC calls (the problem you mentioned previously).  However, the issue cleared up after the minute passed.

----

### Reproduction Case (for at least one of my issues)

1. Launch `parity.exe ui` from command line with RPC logging enabled.
2. Watch log spam for a bit, you'll see several seconds of really fast spam followed by a 1-2 second pause in the spam.  Once the spam resumes, wait 3-4 seconds and then close the UI.
3. Notice that you are now spammed with a bunch of websocket closed errors.
4. Notice that Parity will continue to burn CPU for the remainder of this run.

Basically the trick is to close the UI when it is in the middle of spamming out a ton of a certain class of requests (not sure which exactly, there is a _lot_ of spam).  This will result in a bunch of failures (maybe related to CPU burn, maybe not) and then parity.exe will spend the rest of its run burning CPU.  Re-opening the UI doesn't resolve the problem, though each time I open the UI and close it while it is fetching data memory utilization increases.

While this definitely appears to be exacerbated by the RPC call spam that you mentioned @5chdn, I suspect there is a second issue in here where Parity is leaking threads if the websocket is closed at the wrong time.  These threads, I am guessing, are doing something like sitting in a spin lock waiting for the websocket response (which they will never get).

**Comment by 5chdn** [comment]
I don't have windows available. Didn't know environment variables are not working in config. Thanks for the heads-up. #6417

> Notice that you are now spammed with a bunch of websocket closed errors.

Thanks, this is #6406. What you see is directly related to RPC not catching up in #6387. The client sends old responses to a closed session.

> Basically the trick is to close the UI when it is in the middle of spamming out a ton of a certain class of requests (not sure which exactly, there is a lot of spam). This will result in a bunch of failures (maybe related to CPU burn, maybe not) and then parity.exe will spend the rest of its run burning CPU.

This is #6387 but I can't be sure before you share the logs with RPC tracing enabled. Could you upload a full log to pastie or gist somewhere? Thanks!

**Comment by MicahZoltu** [comment]
Based on my understanding of that issue (which may be mistaken), #6387 is about the volume of calls being made over the RPC when the UI is open.  The symptoms I'm seeing are that if you close the UI in the middle of those calls, Parity will consume significant CPU _indefinitely_ (until it is shut down).  While the volume of calls may make it easier to trigger the bug here, there is a much worse problem which is that Parity appears to be leaking some spinning threads (or something along those lines).  I'm worried that if we wait for #6387 to see if it fixes the issue all it will do is cover up the issue by making it harder to reproduce.  I suspect that #6387 is simply what allowed me to reliably reproduce this problem, but is not the core issue.

These assumptions are based on the fact that after the UI is closed and all RPC calls are stopped (per the logs), Once reproduced, Parity has one thread that runs non stop at top speed until Parity is shutdown.

Gist with debug.log can be found here (note, file is truncated, click link at top to see full thing).

----

When doing a manual profile of what the thread is doing (can only identify kernel calls since I don't have debug symbols) I can see that it is frequently hitting `GetQueuedCompletionStatus` indicating that this thread is likely waiting on some IO.  Presumably, it is waiting for something on the websocket (due to the nature of the repro) which means that something wasn't properly cleaned up when the UI was closed.

**Comment by 5chdn** [comment]
Thanks for sharing the logs! 

Duplicate of #6387

Duplicate of #6406 

Duplicate of #6417

**Comment by MicahZoltu** [comment]
I want to stress how dangerous I think closing this issue as a dupe of #6387 is.  Yes, I am suffering from #6387.  Yes, leveraging the bug defined in #6387 is what allowed me to reproduce the bug described here.

However, I am describing a _different_ bug from #6387 and if the dev team ignores this bug until #6387 is fixed reproducing the bug I referenced here will likely become _much harder_.

I have dealt with this class of bug before on previous projects and not having a repro case can cause the bug to plague the system for years as users continually run into the problem but developers are unable to reproduce.

**The bug I am trying to report is a thread leak, not an n+1 RPC call problem.**

The thread leak I am trying to report is not mentioned anywhere in #6387, #6406 or #6417.  _Please_ re-open this as it is not a duplicate of any of the issues you linked.

**Comment by 5chdn** [comment]
Opening for a 2nd opinion, anyone?

**Comment by 5chdn** [comment]
Does this still happen?

**Comment by MicahZoltu** [comment]
As I mentioned above, by fixing #6387 first all that happened was making it so there are no longer any known reproduction steps for the thread leak.  This does _not_ mean the thread leak is fixed and if it isn't I suspect this issue is going to plague you with random one-off bug reports that are incredibly difficult to reproduce.

Something to try to reproduce this issue would be to open a websocket connection and then spam 100,000 RPC requests over it and then un-gracefully sever the WS connection before receiving all of the responses.  However, it is possible that there are specific RPC requests that need to be made in order to reproduce this, so someone more in-the-know about what Parity UI was doing before #6387 was fixed may have better luck reproducing this now.

**Comment by tomusdrw** [comment]
The leak seems to be related to WS server that we are going to completely replace in 1.9 (or 1.10). We have couple of more issus with current implementation and imho it only makes sense to try to reproduce this on a new server. @5chdn maybe add this to some kind of meta-issue for WebSockets server replacement? (Possibly together with #3544, #6326)

**Comment by 5chdn** [comment]
Replaced by #7032.
